File: java/src/P020/ValidParentheses/ValidParentheses.java
package P020.ValidParentheses;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class ValidParentheses {
    public static boolean isValid(String s) {
        Map<Character, Character> map = new HashMap<Character, Character>() {{
            put(')', '(');
            put(']', '[');
            put('}', '{');
        }};
        LinkedList<Character> stack = new LinkedList<>();

        for (int i = 0; i < s.length(); i++) {
            if (map.containsValue(s.charAt(i))) {
                stack.push(s.charAt(i));
            } else if (map.containsKey(s.charAt(i))) {
                if (stack.size() == 0 || stack.pop() != map.get(s.charAt(i))) {
                    return false;
                }
            } else {
                return false;
            }
        }

        return stack.size() == 0;
    }
}

File: python/P020_Valid_Parentheses/ValidParentheses.py
# coding=utf-8

class Solution(object):
    def isValid(self, s):
        pdict = {')': '(', ']': '[', '}': '{'}
        stack = []
        for x in s:
            if x in pdict.values():
                stack.append(x)
            elif x in pdict.keys():
                if len(stack) == 0 or stack.pop() != pdict[x]:
                    return False
            else:
                return False
        return len(stack) == 0
